### PR TITLE
ini configuration for Cuda 11 build in GenConfig

### PR DESCRIPTION
**Pull request by fryeguy52**
@trilinos/framework 

## Motivation
This will allow the cuda 11 build to run on framework gpu resources through our PR system

## Testing
I ran GenConfig to configure Trilinos and then did make/ctest as usual and got the following on the machine we are intending to run this build on:
```
100% tests passed, 0 tests failed out of 2605

Subproject Time Summary:
Amesos           =  31.07 sec*proc (10 tests)
Amesos2          = 178.43 sec*proc (10 tests)
Anasazi          = 355.48 sec*proc (97 tests)
AztecOO          =  25.35 sec*proc (17 tests)
Belos            = 1132.68 sec*proc (165 tests)
Domi             = 176.56 sec*proc (125 tests)
Epetra           =  59.55 sec*proc (64 tests)
EpetraExt        =  15.49 sec*proc (8 tests)
FEI              =  74.83 sec*proc (42 tests)
Galeri           =   2.67 sec*proc (7 tests)
Ifpack           =  77.67 sec*proc (46 tests)
Intrepid         =  78.50 sec*proc (130 tests)
Intrepid2        = 331.78 sec*proc (290 tests)
Isorropia        =  10.10 sec*proc (6 tests)
Kokkos           = 193.95 sec*proc (48 tests)
KokkosKernels    = 465.99 sec*proc (22 tests)
ML               = 197.95 sec*proc (35 tests)
Panzer           = 195.47 sec*proc (67 tests)
Phalanx          =  37.03 sec*proc (36 tests)
Pike             =   3.57 sec*proc (7 tests)
ROL              = 329.68 sec*proc (186 tests)
RTOp             =  14.52 sec*proc (24 tests)
Rythmos          =  49.44 sec*proc (70 tests)
SEACAS           = 583.89 sec*proc (58 tests)
Sacado           = 112.59 sec*proc (311 tests)
Shards           =   1.19 sec*proc (4 tests)
Stratimikos      =  67.65 sec*proc (37 tests)
Teuchos          = 189.32 sec*proc (139 tests)
Thyra            =  77.86 sec*proc (79 tests)
Tpetra           = 1648.18 sec*proc (283 tests)
Triutils         =   3.79 sec*proc (2 tests)
Xpetra           = 357.10 sec*proc (19 tests)
Zoltan2          = 1246.29 sec*proc (161 tests)

Total Test time (real) = 3653.72 sec

```

### CODE REVIEW

**Review by e10harvey — CHANGES_REQUESTED** [review]

File: packages/framework/ini-files/config-specs.ini [hunk 1]
```diff
@@ -2094,3 +2267,93 @@ opt-set-cmake-var Zoltan2_pamgenMeshAdapterTest_scotch_MPI_4_DISABLE BOOL FORCE
...
+opt-set-cmake-var Tpetra_INST_SERIAL BOOL FORCE : ON
+opt-set-cmake-var Zoltan_ENABLE_Scotch BOOL FORCE : OFF
+
+#Trilinos options
+opt-set-cmake-var Trilinos_ENABLE_DEBUG_SYMBOLS BOOL FORCE : OFF
+opt-set-cmake-var Trilinos_ENABLE_SECONDARY_TESTED_CODE BOOL FORCE : OFF
+
+#Package Enables (for testing jfrye - remove when finished)
+opt-set-cmake-var Trilinos_ENABLE_Kokkos BOOL FORCE : ON
+opt-set-cmake-var Trilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES BOOL FORCE : ON
```
[the remark this answers is not in the document]
Comment (fryeguy52) [reply packages/framework/ini-files/config-specs.ini:2356]: yes. Thanks for catching that

File: packages/framework/ini-files/config-specs.ini [hunk 2]
```diff
@@ -2094,3 +2267,93 @@ opt-set-cmake-var Zoltan2_pamgenMeshAdapterTest_scotch_MPI_4_DISABLE BOOL FORCE
...
+opt-set-cmake-var Phalanx_KOKKOS_DEVICE_TYPE STRING FORCE : CUDA
+opt-set-cmake-var Sacado_ENABLE_HIERARCHICAL_DFAD BOOL FORCE : ON
+opt-set-cmake-var Tpetra_INST_SERIAL BOOL FORCE : ON
+opt-set-cmake-var Zoltan_ENABLE_Scotch BOOL FORCE : OFF
+
+#Trilinos options
+opt-set-cmake-var Trilinos_ENABLE_DEBUG_SYMBOLS BOOL FORCE : OFF
+opt-set-cmake-var Trilinos_ENABLE_SECONDARY_TESTED_CODE BOOL FORCE : OFF
+
+#Package Enables (for testing jfrye - remove when finished)
```
[the remark this answers is not in the document]
Comment (fryeguy52) [reply packages/framework/ini-files/config-specs.ini]: fixed

**Review by e10harvey — CHANGES_REQUESTED** [review]
Looks good. Please remove the changes made to `cmake/std`; related to: https://github.com/trilinos/Trilinos/pull/10648.

File: packages/framework/ini-files/config-specs.ini [hunk 3]
```diff
@@ -2094,3 +2267,89 @@ opt-set-cmake-var Zoltan2_pamgenMeshAdapterTest_scotch_MPI_4_DISABLE BOOL FORCE
...
+opt-set-cmake-var SuperLU_INCLUDE_DIRS STRING FORCE : ${SUPERLU_INC|ENV}
+opt-set-cmake-var SuperLU_LIBRARY_DIRS STRING FORCE : ${SUPERLU_LIB|ENV}
+
+#CXX Settings
+opt-set-cmake-var CMAKE_CXX_STANDARD STRING FORCE : 14
+opt-set-cmake-var CMAKE_CXX_FLAGS STRING : -fPIC -Wall -Warray-bounds -Wchar-subscripts -Wcomment -Wenum-compare -Wformat -Wuninitialized -Wmaybe-uninitialized -Wmain -Wnarrowing -Wnonnull -Wparentheses -Wreorder -Wreturn-type -Wsign-compare -Wsequence-point -Wtrigraphs -Wunused-function -Wunused-but-set-variable -Wunused-variable -Wwrite-strings
+
+#Package Options
+opt-set-cmake-var EpetraExt_ENABLE_HDF5 BOOL FORCE : OFF
+opt-set-cmake-var KOKKOS_ARCH STRING FORCE : VOLTA70
```
Comment (e10harvey) [inline packages/framework/ini-files/config-specs.ini:2336]:
```suggestion
```
`use KOKKOS-ARCH|VOLTA70` already does this.

File: packages/framework/ini-files/config-specs.ini [hunk 4]
```diff
@@ -2094,3 +2267,89 @@ opt-set-cmake-var Zoltan2_pamgenMeshAdapterTest_scotch_MPI_4_DISABLE BOOL FORCE
...
+opt-set-cmake-var KOKKOS_ENABLE_DEBUG BOOL FORCE : ON
```
Comment (e10harvey) [inline packages/framework/ini-files/config-specs.ini:2337]: This is a `_release_` build, should this be removed?

File: packages/framework/ini-files/config-specs.ini [hunk 5]
```diff
@@ -2094,3 +2267,89 @@ opt-set-cmake-var Zoltan2_pamgenMeshAdapterTest_scotch_MPI_4_DISABLE BOOL FORCE
...
+opt-set-cmake-var Kokkos_ENABLE_CUDA BOOL FORCE : ON
+opt-set-cmake-var Kokkos_ENABLE_CUDA_LAMBDA BOOL FORCE : ON
+opt-set-cmake-var Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE BOOL FORCE : OFF
```
Comment (e10harvey) [inline packages/framework/ini-files/config-specs.ini:2340]:
```suggestion
```

This is a `_no-rdc_` build, `use USE-RDC|NO` already does this.

File: packages/framework/ini-files/config-specs.ini [hunk 6]
```diff
@@ -2094,3 +2267,89 @@ opt-set-cmake-var Zoltan2_pamgenMeshAdapterTest_scotch_MPI_4_DISABLE BOOL FORCE
...
+opt-set-cmake-var Kokkos_ENABLE_CXX11_DISPATCH_LAMBDA BOOL FORCE : ON
+opt-set-cmake-var Kokkos_ENABLE_Debug_Bounds_Check BOOL FORCE : ON
```
Comment (e10harvey) [inline packages/framework/ini-files/config-specs.ini:2342]: This is a _release_ build, should this be removed?

File: packages/framework/ini-files/config-specs.ini [hunk 7]
```diff
@@ -2094,3 +2267,89 @@ opt-set-cmake-var Zoltan2_pamgenMeshAdapterTest_scotch_MPI_4_DISABLE BOOL FORCE
...
+opt-set-cmake-var MPI_EXEC_POST_NUMPROCS_FLAGS STRING FORCE : "-map-by;socket:PE=4"
```
Comment (e10harvey) [inline packages/framework/ini-files/config-specs.ini:2343]: This looks like a candidate for a new `[RHEL7-CUDA]` common section.

File: packages/framework/ini-files/config-specs.ini [hunk 8]
```diff
@@ -2094,3 +2267,89 @@ opt-set-cmake-var Zoltan2_pamgenMeshAdapterTest_scotch_MPI_4_DISABLE BOOL FORCE
...
+opt-set-cmake-var Panzer_FADTYPE STRING FORCE : "Sacado::Fad::DFad<RealType>"
+opt-set-cmake-var Phalanx_KOKKOS_DEVICE_TYPE STRING FORCE : CUDA
+opt-set-cmake-var Sacado_ENABLE_HIERARCHICAL_DFAD BOOL FORCE : ON
+opt-set-cmake-var Tpetra_INST_SERIAL BOOL FORCE : ON
+opt-set-cmake-var Zoltan_ENABLE_Scotch BOOL FORCE : OFF
+
+#Trilinos options
+opt-set-cmake-var Trilinos_ENABLE_DEBUG_SYMBOLS BOOL FORCE : OFF
```
Comment (e10harvey) [inline packages/framework/ini-files/config-specs.ini:2351]: Is this necessary (is Trilinos_ENABLE_DEBUG_SYMBOLS set to ON via `use BUILD-TYPE|RELEASE`)?

File: packages/framework/ini-files/config-specs.ini [hunk 9]
```diff
@@ -2094,3 +2267,89 @@ opt-set-cmake-var Zoltan2_pamgenMeshAdapterTest_scotch_MPI_4_DISABLE BOOL FORCE
 opt-set-cmake-var Zoltan2_scotch_example_MPI_4_DISABLE BOOL FORCE : ON
 
 use RHEL7_POST
+
+[rhel7_sems-cuda-11.4.2-sems-gnu-10.1.0-sems-openmpi-4.0.5_release_static_Volta70_no-asan_complex_no-fpic_mpi_no-pt_no-rdc_no-uvm_deprecated-on_no-package-enables]
```
Comment (e10harvey) [inline packages/framework/ini-files/config-specs.ini:2271]:
```suggestion
[rhel7_sems-cuda-11.4.2-sems-gnu-10.1.0-sems-openmpi-4.0.5_release_static_Volta70_no-asan_complex_no-fpic_mpi_pt_no-rdc_no-uvm_deprecated-on_no-package-enables]
```

**Review by ZUUL42 — APPROVED** [review]
I thought these .cmake files were going away as well. But if not, so be it.
One non-functional thing ... you might want to update the version in the comments of the .cmake files. It's still "CUDA 10.1.105".

**Comment by e10harvey** [comment]
> @fryeguy52 - the ctest output you posted above shows only 67 panzer tests enabled for this build. Panzer usually has 200 tests enabled. Could you post the cmake configure output for this build to see why not all panzer tests are enabled?

@rppawlo: This output is available to anyone with access to the trilinos CDash site by searching for "cuda":
- https://trilinos-cdash.sandia.gov/index.php?project=Trilinos&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=cuda
- https://trilinos-cdash.sandia.gov/build/322907/configure

**Comment by rppawlo** [comment]
Thanks @e10harvey ! From the link you posted, it looks like STK and PanzerAdaptersSTK are explicitly being disabled by the test harness in this build. Can you tell me why these packages are disabled in the cuda 11 build? I did not find an open ticket, but maybe missed it? This leaves the majority of Panzer code untested. We really should not rely on this cuda build replacing the vortex cuda build until these packages are enabled. Ifpack2 and MueLu probably also fall into this category too.
```
Explicitly disabled packages on input (by user or by default):  Ifpack2 TriKota STK NOX MueLu ShyLU_DD Stokhos Piro PyTrilinos NewPackage TrilinosCouplings 11
Explicitly disabled SE packages on input (by user or by default):  ShyLU_NodeTacho ShyLU_NodeBasker ShyLU_NodeFastILU SEACASExotec2 Ifpack2 TriKota STK NOX MueLu ShyLU_DDCore ShyLU_DD Stokhos Piro PanzerAdaptersSTK PanzerMiniEM PanzerExprEval PyTrilinos NewPackage TrilinosCouplings 19
```

**Comment by e10harvey** [comment]
> Thanks @e10harvey ! From the link you posted, it looks like STK and PanzerAdaptersSTK are explicitly being disabled by the test harness in this build. Can you tell me why these packages are disabled in the cuda 11 build? I did not find an open ticket, but maybe missed it? This leaves the majority of Panzer code untested. We really should not rely on this cuda build replacing the vortex cuda build until these packages are enabled. Ifpack2 and MueLu probably also fall into this category too.

I checked our tickets for this work but did not find any information about this. We will have to ask the author of:
https://github.com/trilinos/Trilinos/blob/master/packages/framework/ini-files/config-specs.ini#L1618

**Comment by e10harvey** [comment]
CC: @fryeguy52

**Pull request merged**